### PR TITLE
Add DTAgent to AIMA3e

### PR DESCRIPTION
**Pull request by ritwik12**
Solves #37 
Took help from [AIMA4e DT-Agent]( https://github.com/aimacode/aima-java/blob/AIMA4e/core/src/main/java/aima/core/agent/basic/DecisionTheoreticAgent.java) and [DT-Agent Pseudocode](https://github.com/aimacode/aima-pseudocode/blob/master/md/DT-Agent.md#dt-agent)

### CODE REVIEW

File: aima-core/src/main/java/aima/core/agent/DTAgent.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+import java.util.List;
+
+/**
+ * Artificial Intelligence A Modern Approach (3rd Edition): Figure 13.1, page 484.<br>
+ * 
+ * Figure 13.1  A decision-theoretic agent that selects rational actions.
+ * 
+ * @author Ritwik Sharma
+ */
```
Comment (samagra14) [inline aima-core/src/main/java/aima/core/agent/DTAgent.java]: Always add the name of all the authors from whom you have copied the code. In this case, the code is entirely similar to the AIMA4e branch. So, you should give the credit to the person who wrote it (even if you have copied just a single line).

File: aima-core/src/main/java/aima/core/agent/DTAgent.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import aima.core.agent.Percept;
+import aima.core.agent.impl.AbstractAgent;
+import aima.core.probability.BeliefState;
+
+import java.util.List;
+
+/**
+ * Artificial Intelligence A Modern Approach (3rd Edition): Figure 13.1, page 484.<br>
+ * 
+ * Figure 13.1  A decision-theoretic agent that selects rational actions.
```
Comment (samagra14) [inline aima-core/src/main/java/aima/core/agent/DTAgent.java:15]: Pseudocode is missing from the javadoc comments.

File: aima-core/src/main/java/aima/core/agent/DTAgent.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	public Action perceive(Percept percept) {
+
+		// update belief_state based on action and percept
+		belief_state.update(action,percept);
+		/** 
+		 * calculate outcome probabilities for actions,
+	   	 *		given action descriptions and current belief_state 
+		*/	
+		List<Action> action_probabilities = cal_action_probabilities(action_descriptions,belief_state);
+		/**
```
Comment (samagra14) [inline aima-core/src/main/java/aima/core/agent/DTAgent.java]: Where are you gonna store probabilities calculated from this function? You can't store them in the action_probabilities variable as it is a list of actions and hence cannot store probability values.

File: aima-core/src/main/java/aima/core/probability/BeliefState.java [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package aima.core.probability;
+
+public interface BeliefState<Action,Percept> {
+    void update(Action action,Percept percept);
+
+}
```
Comment (samagra14) [inline aima-core/src/main/java/aima/core/probability/BeliefState.java:6]: You don't require this interface for the AIMA3e branch. You can build DTAgent via existing interfaces.

**Review by samagra14 — CHANGES_REQUESTED** [review]

**Comment by ritwik12** [comment]
@samagra14 Thanks for your reviews. I will fix them asap

File: aima-core/src/main/java/aima/core/agent/DTAgent.java (hunk 3 above)
Comment (ritwik12) [reply aima-core/src/main/java/aima/core/agent/DTAgent.java]: That Came to my mind too. Well, one solution to that is your implementation in AIMA4e using aima.core.util.datastructure.Pair. I am thinking of some other solution. Thanks for pointing out the mistakes. @samagra14

File: aima-core/src/main/java/aima/core/probability/BeliefState.java (hunk 4 above)
Comment (ritwik12) [reply aima-core/src/main/java/aima/core/probability/BeliefState.java:6]: I tried to find but can't find any. Also, this gives exclusivity to this algorithm as using BeliefState it gets much closer to the pseudocode.

File: aima-core/src/main/java/aima/core/agent/DTAgent.java (hunk 2 above)
Comment (ritwik12) [reply aima-core/src/main/java/aima/core/agent/DTAgent.java:15]: I've added that, thanks @samagra14

Comment (samagra14) [reply aima-core/src/main/java/aima/core/agent/DTAgent.java:15]: Looks good.

File: aima-core/src/main/java/aima/core/agent/DTAgent.java (hunk 3 above)
Comment (samagra14) [reply aima-core/src/main/java/aima/core/agent/DTAgent.java]: I guess that for this to work you will have to introduce an Action object that stores probabilities. Are you thinking of anything else?

File: aima-core/src/main/java/aima/core/agent/DTAgent.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+		 * calculate outcome probabilities for actions,
+	   	 *		given action descriptions and current belief_state 
+		*/	
+		Action action_probabilities = cal_action_probabilities(action_descriptions,belief_state);
+		/**  
+		 * select action with highest expected utility
+	   	 *		given probabilities of outcomes and utility information
+		 */
+		Action finalAction = action_with_highest_expected_utility(action_probabilities);
+
```
Comment (samagra14) [inline aima-core/src/main/java/aima/core/agent/DTAgent.java:60]: @ritwik12 Here, you are giving this function a single Action. How can the method select an action with the highest utility from a single action? There is nothing to choose from.

File: aima-core/src/main/java/aima/core/agent/DTAgent.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+     */
+	 abstract Action cal_action_probabilities(Action action_description, BeliefState belief_state);
+
+    /**
+     * @param action_probabilities 
+     *			          Probabilities of various outcomes
+     * @return Action with the highest probability.
+     */
+	public abstract Action action_with_highest_expected_utility(Action action_probabilities);
+}
```
Comment (samagra14) [inline aima-core/src/main/java/aima/core/agent/DTAgent.java]: Cannot select an action with the highest utility from a single action.

**Review by samagra14 — CHANGES_REQUESTED** [review]

**Comment by ritwik12** [comment]
@samagra14 I couldn't come up with a unique and better implementation than AIMA4e implementation. you can submit a PR if you have a better implementation for AIMA3e.

**Review by utsavoza — COMMENTED** [review]
I would recommend following the existing coding conventions already used in the project i.e. using CamelCase as a practice for writing names of methods and variables. I think that also applies to the implementation provided in `AIMA4e` branch. And also make sure, the javadoc you provide is properly aligned.

**Comment by ritwik12** [comment]
@norvig Please have a look

**Review by samagra14 — APPROVED** [review]

**Comment by samagra14** [comment]
@norvig This implementation appears correct. Please have a look.

**Comment by ritwik12** [comment]
@norvig Please have a look

**Pull request merged**